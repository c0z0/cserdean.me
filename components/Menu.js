import React from 'react';
import Link from 'next/link';

import { blue, pink } from '../utils/colors';

const items = [
  {
    title: 'Home',
    target: '/',
  },
  {
    title: 'Projects',
    target: '/projects',
  },
  {
    title: 'AI Experiments',
    target: '/ai',
  },
  {
    title: 'CV',
    target: '/static/cv_cosmin_serdean.pdf',
    right: true,
  },
  {
    title: 'GitHub',
    target: 'https://github.com/c0z0',
    right: true,
  },
  {
    title: 'Email',
    target: 'mailto:cosmoserdean@gmail.com',
    right: true,
    primary: true,
  },
];

export default function Navbar({ active = '/' }) {
  const leftItems = items.filter(({ right }) => !right);

  const rightItems = items.filter(({ right }) => right);

  return (
    <div className="menu">
      <div className="menu__left">
        <a href="https://github.com/c0z0/cserdean.me" target="_blank">
          <img src="/static/triangle-gs.svg" alt="logo" className="menu__logo" />
        </a>
        {leftItems.map(({ title, target, primary, right }, k) => (
          <Link href={target} prefetch={!right} key={k}>
            <a
              href={target}
              className={`menu__item ${active === target ? 'menu__item--active' : ''} ${primary
                ? 'menu__item--primary'
                : ''}`}
            >
              {title}
            </a>
          </Link>
        ))}
      </div>
      <div className="menu__right">
        {rightItems.map(({ title, target, primary, right }, k) => (
          <a
            key={k}
            href={target}
            className={`menu__item ${active === target ? 'menu__item--active' : ''} ${primary
              ? 'menu__item--primary'
              : ''}`}
            target="_blank"
          >
            {title}
          </a>
        ))}
      </div>
      <style jsx>{`
        .menu {
          display: flex;
          flex-direction: row;
          align-items: center;
          justify-content: space-between;
          max-width: 900px;
          width: 100%;
          margin: 40px auto;
          font-size: 12px;
          font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell',
            'Fira Sans';
        }

        .menu__left {
          display: flex;
          flex-direction: row;
          align-items: center;
        }

        .menu__item {
          margin: 0 10px;
          color: #888;
          font-size: 12px;
          text-decoration: none;
        }

        .menu__item--active {
          color: #484848 !important;
        }

        .menu__item--primary {
          color: #484848;
          border-radius: 2rem;
          background-color: ${pink};
          padding: 0.625rem 1.5rem;
          color: white;
        }

        .menu__logo {
          height: 40px;
          margin-right: 20px;
          transition: all 0.2s;
          transform: rotate(90deg);
        }

        .menu__logo:hover {
        }

        .menu__dot {
          display: inline-block;
          width: 34px;
          background: ${blue};
          border-radius: 100%;
          margin-right: 12px;
        }
      `}</style>
    </div>
  );
}
