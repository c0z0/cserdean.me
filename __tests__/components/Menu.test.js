import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import Router from "next/router";
import Menu, { items } from "../../components/Menu";

const mockedRouter = {
  push: jest.fn(async () => {}),
  prefetch: () => {}
};
Router.router = mockedRouter;

afterEach(cleanup);

// test("Menu matches snapshot", () => {
//   const { container } = render(<Menu active="/" />);

//   expect(container).toMatchSnapshot();
// });

test("Dark menu render", () => {
  expect(() => {
    render(<Menu dark active="/" />);
  }).not.toThrow();
});

test("Menu sets active page correctly", () => {
  const { container, rerender } = render(<Menu active="/" />);

  items
    .filter(item => !item.right)
    .forEach(item => {
      rerender(<Menu active={item.target} />);

      const activeItem = container.querySelector(".menu__item--active");

      expect(activeItem.getAttribute("href")).toBe(item.target);
      expect(activeItem.textContent).toBe(item.title);
    });
});

test("Menu opens and closes", () => {
  const { container } = render(<Menu active="/" />);

  const menuButton = container.querySelector("button");

  expect(
    container.querySelector(".menu").classList.contains("menu--open")
  ).toBeFalsy();

  fireEvent.click(menuButton);

  expect(
    container.querySelector(".menu").classList.contains("menu--open")
  ).toBeTruthy();

  fireEvent.click(menuButton);

  expect(
    container.querySelector(".menu").classList.contains("menu--open")
  ).toBeFalsy();
});

test("Menu navigates", () => {
  const { container } = render(<Menu active="/" />);

  items
    .filter(item => !item.right)
    .forEach(item => {
      fireEvent.click(
        container.querySelector(`.menu__item[href="${item.target}"]`)
      );

      expect(mockedRouter.push).toBeCalledWith(item.target, item.target, {
        shallow: undefined
      });

      expect(1).toBeGreaterThan(0);

      mockedRouter.push.mockClear();
    });
});

test("Logo right-click navigates to GithubRepo", () => {
  const { container } = render(<Menu active="/" />);

  const logo = container.querySelector(".menu__logo__wrapper *");

  window.location.assign = jest.fn();

  fireEvent.contextMenu(logo);

  expect(window.location.assign).toBeCalledWith(
    "http://github.com/c0z0/cserdean.me"
  );
});
