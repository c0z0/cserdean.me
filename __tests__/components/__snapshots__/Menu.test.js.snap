// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapshot 1`] = `
.c0 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen', 'Ubuntu','Cantarell','Fira Sans',sans-serif;
  background: none;
  padding-top: 20px;
  position: static;
  z-index: 100;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 20px;
  margin-top: 0;
}

.c2 {
  height: 40px;
  width: 47px;
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
  margin-right: 20px;
  -webkit-transition: all 0.5s ease-in-out;
  transition: all 0.5s ease-in-out;
  box-shadow: 0 0 0 0 rgba(0,0,0,0);
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c3 {
  background: none;
  border: none;
  outline: none;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 40px;
  height: 40px;
  padding: 0;
}

.c4 * {
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  width: 22px;
  height: 1px;
  background-color: #484848;
}

.c5 {
  -webkit-transform: rotate(0) translateY(-4px);
  -ms-transform: rotate(0) translateY(-4px);
  transform: rotate(0) translateY(-4px);
}

.c6 {
  -webkit-transform: rotate(0) translateY(4px);
  -ms-transform: rotate(0) translateY(4px);
  transform: rotate(0) translateY(4px);
}

.c7 {
  padding-top: 20px;
  display: none;
}

.c8 {
  color: #484848;
  font-size: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 16px;
  padding-left: 20px;
  border-bottom: 1px solid #eee;
  text-transform: uppercase;
  background: #eee;
  border-color: rgba(0,0,0,0);
}

.c9 {
  color: #484848;
  font-size: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 16px;
  padding-left: 20px;
  border-bottom: 1px solid #eee;
  text-transform: uppercase;
}

.c10 {
  color: #484848;
  font-size: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 16px;
  padding-left: 20px;
  border-bottom: 1px solid #eee;
  text-transform: uppercase;
  color: #FE51BB;
}

@media (min-width:992px) {
  .c0 {
    width: 100%;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    background: none;
  }
}

@media (min-width:992px) {
  .c1 {
    margin: 0;
  }
}

@media (min-width:992px) {
  .c3 {
    display: none;
  }
}

@media (min-width:992px) {
  .c7 {
    padding: 0;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    font-size: 12px;
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
  }
}

@media (min-width:992px) {
  .c8 {
    margin: 0 10px;
    display: inline-block;
    padding: 0;
    color: #999999;
    text-align: left;
    border-bottom: none;
    text-transform: none;
    -webkit-transition: all 0.2s linear;
    transition: all 0.2s linear;
    color: #484848;
    background: none;
  }

  .c8:hover {
    color: #484848;
  }
}

@media (min-width:992px) {
  .c9 {
    margin: 0 10px;
    display: inline-block;
    padding: 0;
    color: #999999;
    text-align: left;
    border-bottom: none;
    text-transform: none;
    -webkit-transition: all 0.2s linear;
    transition: all 0.2s linear;
  }

  .c9:hover {
    color: #484848;
  }
}

@media (min-width:992px) {
  .c10 {
    margin: 0 10px;
    display: inline-block;
    padding: 0;
    color: #999999;
    text-align: left;
    border-bottom: none;
    text-transform: none;
    -webkit-transition: all 0.2s linear;
    transition: all 0.2s linear;
    border-radius: 2rem;
    background-color: #FE51BB;
    padding: 0.625rem 1.5rem;
    color: white !important;
    margin: 0;
    margin-left: 20px;
  }

  .c10:hover {
    color: #484848;
  }
}

<header>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <a
        data-testid="menu-logo"
        href="/"
      >
        <svg
          class="c2"
          height="1220"
          title="source"
          viewBox="0 0 708 610"
          width="1416"
        >
          <path
            d="M301.786 163.964l53.441 95.207-150.893 259.171H707.31L657.013 610H53.441z"
            fill="#c1c1c1"
            fill-rule="evenodd"
          />
          <path
            d="M0 518.342L53.441 610l248.345-439.327 150.893 264.8h100.6L304.929 0z"
            fill="#d0d0d0"
            fill-rule="evenodd"
          />
          <path
            d="M251.64 436.166l-47.84 82.176H708L411.122.54H304.24l247.269 432.465z"
            fill="#b6b6b6"
            fill-rule="evenodd"
          />
        </svg>
      </a>
      <button
        class="c3"
        data-testid="menu-button"
        type="button"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          />
          <div
            class="c6"
          />
        </div>
      </button>
    </div>
    <div
      class="c7"
      data-testid="menu-wrapper"
    >
      <div>
        <a
          class="c8"
          data-testid="menu-item-target-/"
          href="/"
        >
          Home
        </a>
        <a
          class="c9"
          data-testid="menu-item-target-/projects"
          href="/projects"
        >
          Projects
        </a>
        <a
          class="c9"
          data-testid="menu-item-target-/ai"
          href="/ai"
        >
          AI Experiments
        </a>
      </div>
      <div>
        <a
          class="c9"
          data-testid="menu-item-target-/static/cv_cosmin_serdean.pdf"
          href="/static/cv_cosmin_serdean.pdf"
          rel="noopener noreferrer"
          target="_blank"
        >
          CV
        </a>
        <a
          class="c9"
          data-testid="menu-item-target-https://github.com/c0z0"
          href="https://github.com/c0z0"
          rel="noopener noreferrer"
          target="_blank"
        >
          GitHub
        </a>
        <a
          class="c10"
          data-testid="menu-item-target-mailto:cosmin@cserdean.com"
          href="mailto:cosmin@cserdean.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Email
        </a>
      </div>
    </div>
  </div>
</header>
`;
